### PR TITLE
Option to delay catalog initialization / rebind until after required OSGi services are available

### PR DESCRIPTION
**Pull request by ahgittin**
More details will be in a docs PR which will come shortly and be advertised here

### CODE REVIEW

File: karaf/init/src/main/java/org/apache/brooklyn/launcher/osgi/OsgiLauncherImpl.java [hunk 1]
```diff
@@ -108,21 +136,178 @@ public void initOsgi() {
...
+
+        Framework f = (Framework) bundle.getBundleContext().getBundle(0);
+        int startLevel = f.adapt(FrameworkStartLevel.class).getStartLevel();
+
+        if (areServiceDependenciesReady(bundle, bundle + " bundle activation")) {
+            LOG.debug("Starting OSGi catalog/rebind (no service dependencies or all already satisfied, on bundle activation)");
+            doStartOsgiAfterBundlesRefreshed();
+
+        } else {
+            ServiceListener sl[] = { null };
```
Comment (duncangrant) [inline karaf/init/src/main/java/org/apache/brooklyn/launcher/osgi/OsgiLauncherImpl.java:152]: Very minor but not the standard java way.

File: karaf/init/src/main/java/org/apache/brooklyn/launcher/osgi/OsgiLauncherImpl.java [hunk 2]
```diff
@@ -108,21 +136,178 @@ public void initOsgi() {
...
+            Duration timeout;
+            try {
+                timeout = Duration.parse((String) getBrooklynProperties().getConfig(BROOKLYN_OSGI_DEPENDENCIES_SERVICES_TIMEOUT));
+            } catch (Exception e) {
+                throw Exceptions.propagateAnnotated("Invalid duration specified for '"+BROOKLYN_OSGI_DEPENDENCIES_SERVICES_TIMEOUT+"'", e);
+            }
+            fallbackThread = new Thread(() -> {
+                if (timeout==null) {
+                    LOG.debug("No timeout specified for '"+BROOKLYN_OSGI_DEPENDENCIES_SERVICES_TIMEOUT+"'; will wait indefinitely for service dependencies");
+                    return;
```
Comment (duncangrant) [inline karaf/init/src/main/java/org/apache/brooklyn/launcher/osgi/OsgiLauncherImpl.java:181]: I don't think you want to return here?

**Review by duncangrant — APPROVED** [review]
LGTM thanks @ahgittin

File: karaf/init/src/main/java/org/apache/brooklyn/launcher/osgi/OsgiLauncherImpl.java (hunk 2 above)
Comment (ahgittin) [reply karaf/init/src/main/java/org/apache/brooklyn/launcher/osgi/OsgiLauncherImpl.java]: good spot - thx, fixed

**Pull request merged**